### PR TITLE
VS2015 compatibility

### PR DESCRIPTION
**Pull request by gnieboer**
Includes project files, an AVX2-only option, ml compatibility, and snprintf inclusion

### CODE REVIEW

**Comment by mlt** [comment]
One can generate any VS project files with CMake already supported.

**Comment by gnieboer** [comment]
True but the repo has pre-generated files for other VS versions, so this would just keep what is already there up to date.  Not all users are familiar with CMake.

File: contrib/masmx86/bld_ml32.bat [hunk 1]
```diff
@@ -1,2 +1,2 @@
-ml /coff /Zi /c /Flmatch686.lst match686.asm
-ml /coff /Zi /c /Flinffas32.lst inffas32.asm
+ml /coff /Zi /c /safeseh /Flmatch686.lst match686.asm
+ml /coff /Zi /c /safeseh /Flinffas32.lst inffas32.asm
```
Comment (mtl1979) [inline contrib/masmx86/bld_ml32.bat]: What is the reason for adding /safeseh for older versions of Visual Studio... Is it actually supported and required in all versions?

Comment (gnieboer) [inline contrib/masmx86/bld_ml32.bat]: Older ml.exe versions back to VS 2003.NET support /safeseh.  From MS: "Marks the object as either containing no exception handlers or containing exception handlers that are all declared with .SAFESEH"

So if the .asm files are build w/o SAFESEH, and the VS linker flag /SAFESEH:YES is set, then the build will fail.  The Linker flag YES is now the default in VS 2012+.  Building the .asm files with safeseh has no effect on earlier VS version where the linker is not defaulting to YES.

Having said all that, since there is a VS2015 specific asm build file being added, there is no danger in discarding this change if desired.  I could only test on VS2015 so I cannot validate the regression.

Comment (mtl1979) [inline contrib/masmx86/bld_ml32.bat]: If it is recommended for older versions, I would think it should be separated as different patch set, so it can be validated separately by people who still use older Visual Studio version. The other choice would be dirty hack in my opinion as it would combine the two batch files and add detection for Visual Studio version and select code paths depending on it.

**Comment by JBlond** [comment]
To have a save option for all VC version there is just a single line to be changed

zlib\win32\Makefile.msc 
Just line 31

`ASFLAGS = -coff -Zi -safeseh $(LOC)`

copy zlib\win32\zlib.def and zlib\win32\zlib1.rc into zlib
For x64
`nmake -f win32/Makefile.msc AS=ml64 LOC="-DASMV -DASMINF -I." OBJA="inffasx64.obj gvmat64.obj inffas8664.obj`

x86
`nmake -f win32/Makefile.msc LOC="-DASMV -DASMINF" OBJA="inffas32.obj match686.obj"`

**Comment by madler** [comment]
There are many Visual Studio pull requests. Can someone who is able to test them across versions of Visual Studio merge them into a single pull request? Also please test it with the latest zlib develop branch here on github, and account for functions that have been added since 1.2.8. Thank you.

**Comment by AraHaan** [comment]
See #182 which is based from #167. Other than that it also Provides some other fixes to the definition file and the resource script for the version info resource. The export Definition file was also fixed in it for all of the Project files and includes VS2013 for those who still use it. It targets the latest Develop branch though.

**Comment by AraHaan** [comment]
Also #182 disables ``SAFESEH`` for VS2013 and VS2015.

**Comment by madler** [comment]
Please see zlib 1.2.9 for updated vstudio support. Please create new issues or new pull requests if there are problems or desired behaviors against that version. Thanks.
